<a href="https://colab.research.google.com/github/OhOmah/DS-Unit-1-Sprint-2-Statistics/blob/master/module1/LS_DS_121_Statistics_Probability_Assignment.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [2]:
### YOUR WORK HERE
# Importing the libaries I need to complete this asignment 
import pandas as pd 
import matplotlib.pyplot as plt 
import numpy as np
from scipy import stats 
# Using a shell command to grabe the automobile dataset 

!wget 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'

--2020-07-14 03:49:25--  https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 25936 (25K) [application/x-httpd-php]
Saving to: ‘imports-85.data’

imports-85.data     100%[===================>]  25.33K  --.-KB/s    in 0.06s   

2020-07-14 03:49:26 (401 KB/s) - ‘imports-85.data’ saved [25936/25936]



In [4]:
# Now assigning the dataset to a variable and checking the header 

df = pd.read_csv('imports-85.data')

df.head()

3    ?  alfa-romero  gas  std   two  ...  9.00  111  5000  21  27  13495
0  3    ?  alfa-romero  gas  std   two  ...   9.0  111  5000  21  27  16500
1  1    ?  alfa-romero  gas  std   two  ...   9.0  154  5000  19  26  16500
2  2  164         audi  gas  std  four  ...  10.0  102  5500  24  30  13950
3  2  164         audi  gas  std  four  ...   8.0  115  5500  18  22  17450
4  2    ?         audi  gas  std   two  ...   8.5  110  5500  19  25  15250

[5 rows x 26 columns]

In [5]:
# Need to fix the column, and also replace the '?' with NaN values 

column_names = ('symboling', 'normalized-losses', 'make', 'fuel-type', 'asperation', 'num-of-doors', 'body-style',
                'drive-wheels', 'engine-location', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'engine-type', 
                'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ration', 'horsepower', 
                'peak-rpm', 'city-mpg', 'highway-mpg', 'price')

df = pd.read_csv('imports-85.data', names=column_names, na_values='?')

df.head()

symboling  normalized-losses         make  ... city-mpg highway-mpg    price
0          3                NaN  alfa-romero  ...       21          27  13495.0
1          3                NaN  alfa-romero  ...       21          27  16500.0
2          1                NaN  alfa-romero  ...       19          26  16500.0
3          2              164.0         audi  ...       24          30  13950.0
4          2              164.0         audi  ...       18          22  17450.0

[5 rows x 26 columns]

In [6]:
# Now since that is done I'm going to check how many and where the nan values are so I can go ahead and exclude them for when run t-tests through the data frame 

df.isnull().sum()

symboling              0
normalized-losses     41
make                   0
fuel-type              0
asperation             0
num-of-doors           2
body-style             0
drive-wheels           0
engine-location        0
wheel-base             0
length                 0
width                  0
height                 0
curb-weight            0
engine-type            0
num-of-cylinders       0
engine-size            0
fuel-system            0
bore                   4
stroke                 4
compression-ration     0
horsepower             2
peak-rpm               2
city-mpg               0
highway-mpg            0
price                  4
dtype: int64

In [7]:
# Grabbing the descriptive statistics of the DF to get a general understanding 

df.describe()

symboling  normalized-losses  ...  highway-mpg         price
count  205.000000         164.000000  ...   205.000000    201.000000
mean     0.834146         122.000000  ...    30.751220  13207.129353
std      1.245307          35.442168  ...     6.886443   7947.066342
min     -2.000000          65.000000  ...    16.000000   5118.000000
25%      0.000000          94.000000  ...    25.000000   7775.000000
50%      1.000000         115.000000  ...    30.000000  10295.000000
75%      2.000000         150.000000  ...    34.000000  16500.000000
max      3.000000         256.000000  ...    54.000000  45400.000000

[8 rows x 16 columns]

In [13]:
df.describe(exclude='number')

make fuel-type asperation  ... engine-type num-of-cylinders fuel-system
count      205       205        205  ...         205              205         205
unique      22         2          2  ...           7                7           8
top     toyota       gas        std  ...         ohc             four        mpfi
freq        32       185        168  ...         148              159          94

[4 rows x 10 columns]

In [9]:
# Looking at the shape to know how many vehicles we are working with 
df.shape

(205, 26)

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [21]:
### YOUR WORK HERE
# Grabbing 10 samples of vehicles to work with
sample = df.sample(n=10, random_state=30)

sample.head(10)

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0
103          0              108.0  ...          25  13499.0
26           1              148.0  ...          38   7609.0
20           0               81.0  ...          43   6575.0
155          0               91.0  ...          32   8778.0
185          2               94.0  ...          34   8195.0

[10 rows x 26 columns]

In [23]:
# The mean of the highway mpg of the original df

df['highway-mpg'].mean()

30.75121951219512

In [22]:
# Grabbing the sample mean to compare with the mean of the full data frame. 
sample['highway-mpg'].mean()

32.8

**The sample mean for the 10 random cars is 32.8 while the avg mean of the full population is 30.75, I would say they are different.** 

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

##  1) Null Hypothesis: 
The average highway MPG of the cars for sale is 35 mpg

$H_0: \overline{x}_1 == \overline{x}_2$

## 2) Alternative Hypothesis

The average highway MPG of the cars for sale is **NOT** 35 MPG 

$H_0: \overline{x}_1 \neq \overline{x}_2$

## 3) Confidence Level: 95% 

In [30]:
### YOUR WORK HERE

# Running a T-test to see if his claim is correct 
stats.ttest_1samp(sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

## 4) Conclusion

With a t-statistic of -0.95 and a pvalue of 0.36 we **DO NOT REJECT** the salesman's initial claim that the average mpg is 35. 

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [32]:
### YOUR WORK HERE

# Upgrading the sample size to 100
sample1 = df.sample(n=100, random_state=30)

sample1.shape
sample1.head()

symboling  normalized-losses  ... highway-mpg    price
121          1              154.0  ...          38   6692.0
176         -1               65.0  ...          32  10898.0
24           1              148.0  ...          38   6229.0
88          -1              137.0  ...          30   9279.0
72           3              142.0  ...          18  35056.0

[5 rows x 26 columns]

##  1) Null Hypothesis: 
The average highway MPG of the cars for sale is 35 mpg

$H_0: \overline{x}_1 == \overline{x}_2$

## 2) Alternative Hypothesis

The average highway MPG of the cars for sale is **NOT** 35 MPG 

$H_0: \overline{x}_1 \neq \overline{x}_2$

## 3) Confidence Level: 95%

In [33]:
# Now running a t_test to see the result 
stats.ttest_1samp(sample1['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

## 4) Conclusion

Due to a t-statistic of -6.51 and a p-value of -0 we are **REJECTING** the salesman's claim that the avg highway mpg of the vehicles is 35mpg

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

Since the first t_test was working with a smaller dataset it was much harder to know within reasonable doubt that we can reject it. With a larger dataset it is easier to get more accurate data. 

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### YOUR WORK HERE

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE

### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE